selenium-tests/test_registration.py

```diff
@@ -0,0 +1,152 @@
+import pytest
+from selenium.webdriver.support.select import Select
+import uuid
+
+# TODO: eventually, make this configurable
+EVENT_NAME = "magstock"
+
+@pytest.fixture
+def selenium(selenium):
+    selenium.implicitly_wait(10)
+    selenium.maximize_window()
+    return selenium
+
+
+def ensure_text_in_error_message(driver, text):
+    html = driver.find_elements_by_class_name("toast-message")[0].get_attribute("innerHTML")
+    assert text in html
+
+
+def ensure_text_exists_in_head(driver, text):
+    head_html = driver.find_element_by_tag_name("head").get_attribute("innerHTML")
+    assert text in head_html
+
+
+class ArgsBase:
+    def __init__(self, **kwargs):
+        for key, value in kwargs.items():
+            setattr(self, key, value)
+
+
+class FormValidator(ArgsBase):
+    def __init__(self, **kwargs):
+        self.field_name = None
+        self.field_value = None
+        self.expected_error = None
+        self.event_specific = None
+
+        super(FormValidator, self).__init__(**kwargs)
+
+    def run(self, driver):
+        if self.expected_error:
+            ensure_text_in_error_message(driver, self.expected_error)
+
+        if self.field_name and (not self.event_specific or self.event_specific == EVENT_NAME):
+            elem = self.fill(driver)
+            elem.submit()
+
+    def fill(self, driver):
+        """
+        Fill out a field. Can be overridden, and must return an element that is part of the form
+        that we are filling out.
+        """
+        elem = driver.find_element_by_xpath("//input[@name='{}']".format(self.field_name))
+        if self.field_value:
+            elem.clear()
+            elem.send_keys(self.field_value)
+        return elem
+
+
+class BirthdateFieldValidator(FormValidator):
+    def fill(self, driver):
+        date_values = self.field_value.split('-')
+        assert len(self.field_name) == len(date_values)
+        fields_and_values = zip(self.field_name, date_values)
+
+        elem = None
+        for field, text in fields_and_values:
+            elem = driver.find_elements_by_class_name(field)[0]
+            elem.send_keys(text)
+        return elem
+
+
+class TextFieldValidator(FormValidator):
+    pass
+
+
+class DropDownValidator(TextFieldValidator):
+    def fill(self, driver):
+        elem = driver.find_element_by_name(self.field_name)
+        if self.field_value:
+            Select(elem).select_by_visible_text(self.field_value)
+        return elem
+
+
+class EnsurePage(ArgsBase):
+    def run(self, driver):
+        assert self.expected_page in driver.current_url
+
+
+def generate_unique_email():
```

Comment:
For these generators I highly recommend the Chance library.. using it for test data generation in the Javascript version (/casperjs of this repo) and here is a Python version of the lib: https://github.com/davitv/chance